File: scripts/plotutils.py
# %%
import plotly.express as px
import pandas as pd
import numpy as np
import plotly.io as pio
import os

pio.templates.default = "plotly_white"
if os.environ["QUARTO_FIG_FORMAT"] == "pdf":
    pio.renderers.default = "pdf"

def plot2d(X, y, manifold_method):
    fig = px.scatter(
        x=X[:, 0],  # type: ignore
        y=X[:, 1],  # type: ignore
        color=y,
        category_orders={"color": y.cat.categories},
    )
    fig.update_xaxes(visible=False)
    fig.update_yaxes(visible=False)
    fig.update_layout(legend={"itemsizing": "constant"})
    return fig


def plot3d(X, y, manifold_method):
    fig = px.scatter_3d(
        x=X[:, 0],  # type: ignore
        y=X[:, 1],  # type: ignore
        z=X[:, 2],  # type: ignore
        color=y,  # type: ignore
        category_orders={"color": y.cat.categories},
    )
    fig.update_traces(
        marker=dict(size=2, opacity=0.5),
        selector=dict(mode="markers")
    )
    fig.update_scenes(
        xaxis_visible=False,
        yaxis_visible=False,
        zaxis_visible=False
    )
    fig.update_layout(legend={"itemsizing": "constant"})
    return fig


methods = ["tsne", "sammon", "isomap", "lle"]


def plot_dataset(datasetname, methods=methods, sample_size=None):
    pandas_df = pd.read_csv(f"{datasetname}.csv", header=[0, 1])
    if sample_size is not None:
        pandas_df = pandas_df.sample(sample_size)
    y = pandas_df["y"].iloc[:, 0].astype("category") # type: ignore
    X = pandas_df.drop(columns=["y"])

    for method in methods:
        if X[method].shape[1] == 2:
            plot2d(X[method].values, y, method).show()
        elif X[method].shape[1] == 3:
            plot3d(X[method].values, y, method).show()

File: scripts/plotutils.R
library(plotly)
library(readr)
library(dplyr)

plot2d <- function(X, y, manifold_method) {
    fig <- plot_ly() %>%
        add_trace(x = X[, 1], y = X[, 2], color = y, colors = "Paired", type = "scatter", mode = "markers") %>%
        layout(xaxis = list(visible = FALSE), yaxis = list(visible = FALSE), legend = list(itemsizing = "constant"))
    return(fig)
}

plot3d <- function(X, y, manifold_method) {
    fig <- plot_ly() %>%
        add_trace(x = X[, 1], y = X[, 2], z = X[, 3], color = y, colors = "Paired", type="scatter3d", mode = "markers", marker = list(size = 3, opacity = 0.5)) %>%
        layout(scene = list(xaxis = list(visible = FALSE), yaxis = list(visible = FALSE), zaxis = list(visible = FALSE)), legend = list(itemsizing = "constant"))
    return(fig)
}

plot_dataset <- function(datasetname, methods, sample_size = NULL) {
    headers <- read_csv(paste0(datasetname, ".csv"), col_names = FALSE, show_col_types = FALSE, n_max = 2)
    col_names <- paste(headers[1, ], as.numeric(headers[2, ]) + 1, sep = "_")

    pandas_df <- read_csv(paste0(datasetname, ".csv"), col_names = col_names, show_col_types = FALSE, skip = 2)
    
    if (!is.null(sample_size)) {
        pandas_df <- pandas_df[sample(1:nrow(pandas_df), sample_size), ]
    }
    y <- as.factor(pandas_df$y_1)
    X <- pandas_df[, -ncol(pandas_df)]
    
    figlist <- list(1:length(methods))   
    i <- 1
    for (method in methods) {
        Xm <- X %>% select(starts_with(paste0(method,"_"))) %>% as.matrix
        ncols <- ncol(Xm)
        if (ncols == 2) {
            figlist[[i]] <- plot2d(Xm, y, method)
        } else if (ncols == 3) {
            figlist[[i]] <- plot3d(Xm, y, method)
        }
        i <- i + 1
    }
    return(figlist)
}
